test/TestUtils/TestProjectFixture.cs

```diff
@@ -0,0 +1,290 @@
+
+            var buildArgs = new List<string>();
+            if (runtime != null)
+            {
+                buildArgs.Add("--runtime");
+                buildArgs.Add(runtime);
+            }
+
+            if (framework != null)
+            {
+                buildArgs.Add("--framework");
+                buildArgs.Add(framework);
+            }
+
+            if (outputDirectory != null)
+            {
+                buildArgs.Add("-o");
+                buildArgs.Add(outputDirectory);
+            }
+
+            dotnet.Build(buildArgs.ToArray())
+                .WorkingDirectory(_testProject.ProjectDirectory)
+                .CaptureStdErr()
+                .CaptureStdOut()
+                .Execute()
+                .EnsureSuccessful();
+
+            _testProject.LoadOutputFiles();
+
+            return this;
+        }
+
+        public TestProjectFixture PublishProject(
+            DotNetCli dotnet = null,
+            string runtime = null,
+            string framework = "netcoreapp1.0",
+            string outputDirectory = null)
+        {
+            dotnet = dotnet ?? _sdkDotnet;
+            outputDirectory = outputDirectory ?? _testProject.OutputDirectory;
+            _testProject.OutputDirectory = outputDirectory;
+
+            var publishArgs = new List<string>();
+            if (runtime != null)
+            {
+                publishArgs.Add("--runtime");
+                publishArgs.Add(runtime);
+            }
+
+            if (framework != null)
+            {
+                publishArgs.Add("--framework");
+                publishArgs.Add(framework);
+            }
+
+            if (outputDirectory != null)
+            {
+                publishArgs.Add("-o");
+                publishArgs.Add(outputDirectory);
+            }
+
+            dotnet.Publish(publishArgs.ToArray())
+                .WorkingDirectory(_testProject.ProjectDirectory)
+                .CaptureStdErr()
+                .CaptureStdOut()
+                .Execute()
+                .EnsureSuccessful();
+
+            _testProject.LoadOutputFiles();
+
+            return this;
+        }
+
+        public TestProjectFixture RestoreProject(DotNetCli dotnet = null)
+        {
+            dotnet = dotnet ?? _sdkDotnet;
+
+            dotnet.Restore()
+                .WorkingDirectory(_testProject.ProjectDirectory)
+                .Execute()
+                .EnsureSuccessful();
+
+            return this;
+        }
+
+        public TestProjectFixture EnsureRestored(DotNetCli dotnet = null)
+        {
+            dotnet = dotnet ?? _sdkDotnet;
+
+            if ( ! _testProject.IsRestored())
+            {
+                RestoreProject(dotnet: dotnet);
+            }
+
+            return this;
+        }
+
+        public TestProjectFixture Copy()
+        {
+            return new TestProjectFixture(this);
+        }
+
+        public TestProjectFixture MoveDepsJsonToSubdirectory()
```

Comment:
I'm not excited about the methods here and below. It feels like these "test utils" are duplicating what the product should be doing.  And as we add more artifacts to our product (or refactor the current ones) these methods will have to change. Also, doing this kind of work in tests can easily mask bugs in the product.